train_gpt2.cu

```diff
@@ -581,123 +597,213 @@ void gpt2_forward(GPT2 *model, const int* inputs, const int* targets, size_t B,
     }
+    cudaCheck(cudaEventRecord(input_done, input_copy_stream));
 
-    // validate inputs, all indices must be in the range [0, V)
-    // we can do this while the copies are already underway
-    for(int i = 0; i < B * T; i++) {
-        assert(0 <= inputs[i] && inputs[i] < V);
-        if (targets != NULL) {
+    if (targets != NULL) {
+        cudaCheck(cudaMemcpyAsync(model->targets, targets, B * T * sizeof(int), cudaMemcpyHostToDevice, target_copy_stream));
+        #pragma omp parallel for
+        for (size_t i = 0; i < B * T; i++) {
             assert(0 <= targets[i] && targets[i] < V);
         }
+        cudaCheck(cudaEventRecord(target_done, target_copy_stream));
     }
 
-    // forward pass
-    ParameterTensors params = model->params; // for brevity
+    // Wait for data transfer to complete before starting computation
+    cudaCheck(cudaStreamWaitEvent(compute_stream, input_done, 0));
+    if (targets != NULL) {
+        cudaCheck(cudaStreamWaitEvent(compute_stream, target_done, 0));
+    }
+
+    ParameterTensors params = model->params;
     ActivationTensors acts = model->acts;
-    encoder_forward(acts.encoded, model->inputs, params.wte, params.wpe, B, T, C, main_stream); // encoding goes into residual[0]
+    encoder_forward(acts.encoded, model->inputs, params.wte, params.wpe, B, T, C, compute_stream);
+
+    layernorm_forward((model->recompute < 2) ? acts.ln1 : acts.lnf, acts.ln1_mean, acts.ln1_rstd, acts.encoded, params.ln1w, params.ln1b, B, T, C, compute_stream);
+
+    // Offsets for QKV and attention projection weights and biases
+    size_t qkvw_offset = 3 * C * C;
+    size_t qkvb_offset = 3 * C;
+    size_t attprojw_offset = C * C;
+    size_t attprojb_offset = C;
+
+    // Offsets for layer normalization weights and biases
+    size_t ln2w_offset = C;
+    size_t ln2b_offset = C;
+
+    // Offsets for fully connected weights and biases
+    size_t fcw_offset = 4 * C * C;
+    size_t fcb_offset = 4 * C;
+    size_t fcprojw_offset = C * 4 * C;
+    size_t fcprojb_offset = C;
+
+    // Offsets for activations
+    size_t ln1_offset = B * T * C;
+    size_t qkvr_offset = B * T * 3 * C;
+    size_t atty_offset = B * T * C;
+    size_t attproj_offset = B * T * C;
+    size_t residual2_offset = B * T * C;
+    size_t ln2_offset = B * T * C;
+    size_t ln2_mean_offset = B * T;
+    size_t ln2_rstd_offset = B * T;
+    size_t fch_offset = B * T * 4 * C;
+    size_t fch_gelu_offset = B * T * 4 * C;
+    size_t fcproj_offset = B * T * C;
+    size_t residual3_offset = B * T * C;
+
+    // Offsets for layer normalization recomputation
+    size_t ln1_mean_offset = B * T;
+    size_t ln1_rstd_offset = B * T;
+    size_t ln1w_offset = C;
+    size_t ln1b_offset = C;
+
+    // Offsets for CUDNN and non-CUDNN cases
+    #ifdef ENABLE_CUDNN
+    size_t att_offset = B * NH * T;
+    #else
+    size_t att_offset_noncudnn = B * NH * T * T;
+    #endif
+
+    // prefetch the offsets into cache
```

Comment:
have you profiles this optimization on its own?
I would be **extremely** surprised if this had any effect whatsoever. None of these offsets should be in caches, they should never leave processor registers. 
The compiler should also pretty much figure out the optimal order for these computations, so that offsets that are needed only for later kernels will only be computed when they are needed.